### PR TITLE
Feature/added cli types

### PR DESCRIPTION
**Pull request by taranek**
**What kind of change does this PR introduce?**
Adds Typescript support for `webpack-cli` package. 

**Did you add tests for your changes?**
No

**If relevant, did you update the documentation?**
No

**Summary**
Inspired by [PR #3076](https://github.com/webpack/webpack-cli/pull/3076) I managed to propose my solution with reduced usage of `any`. It it still missing some types from `webpack`(v5.68.0) package (which are declared there but not exported) - for now I duplicated them in `webpack-internal-types.ts` - I can open a seperate PR in `webpack` repo in order to export them (please tell me if it's needed). It's my first PR here, so I had to reverse-engineer some types and names for them - which means I probably didn't pick the best names for types.

**Does this PR introduce a breaking change?**
No

**Other information**
N/A

### CODE REVIEW

File: packages/webpack-cli/src/webpack-cli.ts
Comment (alexander-akait) [inline packages/webpack-cli/src/webpack-cli.ts:2387]: It should never return `undefined`, we should refactor logic

File: packages/webpack-cli/src/webpack-internal-types.ts
Comment (alexander-akait) [inline packages/webpack-cli/src/webpack-internal-types.ts:176]: We should fix it in webpack and export them, no need to duplicate, it is just mistake on webpack side

File: packages/webpack-cli/src/webpack-cli.ts
Comment (alexander-akait) [inline packages/webpack-cli/src/webpack-cli.ts:2496]: Same here, only `boolean` should be returned, technically `undefined` is `false`, but it is not good practice

**Review by alexander-akait — COMMENTED** [review]
Great work! Let's fix some stuff and we can merge

File: packages/webpack-cli/src/webpack-internal-types.ts
Comment (taranek) [reply packages/webpack-cli/src/webpack-internal-types.ts]: Thank you, I've submitted a [PR](https://github.com/webpack/webpack/pull/15377)

File: packages/webpack-cli/src/webpack-cli.ts
Comment (taranek) [reply packages/webpack-cli/src/webpack-cli.ts]: Thanks, I'll just cast both method using `Boolean()`

**Review by alexander-akait — COMMENTED** [review]
Good job, currently we use `any` https://github.com/webpack/webpack-cli/blob/master/packages/info/src/index.ts#L14 on CLI packages, can you fix it to check we really don't missing something in our API (some things can be complex, put TODO in this case)

File: packages/webpack-cli/src/webpack-internal-types.ts
Comment (snitin315) [reply packages/webpack-cli/src/webpack-internal-types.ts]: I believe we need to update webpack here.

**Comment by taranek** [comment]
@alexander-akait Thank you for mentioning that part. I managed to import the types from `webpack-cli` by using project references. I guess we can run CI checks once again.

File: .eslintignore
Comment (snitin315) [inline .eslintignore:5]: This will now ignore `packages/webpack-cli/lib` too which we don't want.

**Comment by snitin315** [comment]
@taranek  Let's fix lint

File: packages/webpack-cli/package.json
Comment (taranek) [inline packages/webpack-cli/package.json:65]: I've added `webpack-dev-server` as dev dep in `webpack-cli` package, but the smoketests are still failing when resolving `webpack-dev-server` in `serve` package. [(check the details here)](https://github.com/webpack/webpack-cli/runs/5214548107?check_suite_focus=true) 

Should I add `webpack-dev-server` as a peerDependency in `serve` package? 🤔

Comment (alexander-akait) [reply packages/webpack-cli/package.json:65]: In theory yes, but there is the problem, `serve` is part of `webpack-cli`, so if we add `webpack-dev-server` to `peerDependency` new npm verison will install it automatically, it is not good, because many projects don't use dev server, I think better refactor our tests

Comment (taranek) [reply packages/webpack-cli/package.json]: Thank you for the explanation. I managed to reproduce the CI error locally and fix it by removing the dev dep from `webpack-cli` and surpressing extraneous import in TS and eslint.

**Review by alexander-akait — APPROVED** [review]
Looks good, let's wait CI

**Review by alexander-akait — DISMISSED** [review]
/cc @webpack/cli-team Can anyone review this as well?

**Review by snitin315 — APPROVED** [review]
This is great ⭐ , thank you ❤️

**Review by rishabh3112 — APPROVED** [review]

**Comment by alexander-akait** [comment]
@taranek Sorry for this, can you rebase?

**Comment by taranek** [comment]
@alexander-akait  Sure, but are we ready to merge now? If we still need a review from the core team, can we rebase after the approval?

**Comment by snitin315** [comment]
Yes, we can merge after rebase.

**Review by rishabh3112 — APPROVED** [review]

**Comment by alexander-akait** [comment]
Big thanks

**Comment by taranek** [comment]
@alexander-akait You're welcome - do you think converting tests code to TS would be worth doing as well?

**Pull request merged**
